include/slb/functional.hpp

```diff
@@ -0,0 +1,195 @@
+  template<class R, class T>
+    unspecified mem_fn(R T::*) noexcept;
+
+  // [func.wrap], polymorphic function wrappers
+  class bad_function_call;
+
+  template<class> class function; // not defined
+  template<class R, class... ArgTypes> class function<R(ArgTypes...)>;
+
+  template<class R, class... ArgTypes>
+    void swap(function<R(ArgTypes...)>&, function<R(ArgTypes...)>&) noexcept;
+
+  template<class R, class... ArgTypes>
+    bool operator==(const function<R(ArgTypes...)>&, nullptr_t) noexcept;
+  template<class R, class... ArgTypes>
+    bool operator==(nullptr_t, const function<R(ArgTypes...)>&) noexcept;
+  template<class R, class... ArgTypes>
+    bool operator!=(const function<R(ArgTypes...)>&, nullptr_t) noexcept;
+  template<class R, class... ArgTypes>
+    bool operator!=(nullptr_t, const function<R(ArgTypes...)>&) noexcept;
+
+  // [func.search], searchers
+  template<class ForwardIterator, class BinaryPredicate = equal_to<>>
+    class default_searcher;
+
+  template<class RandomAccessIterator,
+           class Hash = hash<
+               typename iterator_traits<RandomAccessIterator>::value_type>,
+           class BinaryPredicate = equal_to<>>
+    class boyer_moore_searcher;
+
+  template<class RandomAccessIterator,
+           class Hash = hash<
+               typename iterator_traits<RandomAccessIterator>::value_type>,
+           class BinaryPredicate = equal_to<>>
+    class boyer_moore_horspool_searcher;
+
+  // [unord.hash], hash function primary template
+  template <class T>
+    struct hash;
+
+  // [func.bind], function object binders
+  template <class T>
+    inline constexpr bool is_bind_expression_v = is_bind_expression<T>::value;
+  template <class T>
+    inline constexpr int is_placeholder_v = is_placeholder<T>::value;
+}
+
+*/
+
+#include <functional>
+#include <utility>
+
+#include "detail/config.hpp"
+#include "detail/invoke.hpp"
+#include "type_traits.hpp"
+
+namespace slb {
+
+// [func.invoke], invoke
+
+// We only enable the C++17 implementation under C++2a here to account for
+// P0704: "Fixing const-qualified pointers to members".
+
+#if __cpp_lib_invoke /* C++17 */ && __cplusplus > 201703L /* C++2a */
+using std::invoke;
+#else
+template <typename F, typename... Args>
+typename slb::invoke_result<F, Args...>::type
+invoke(F&& f,
+       Args&&... args) noexcept(slb::is_nothrow_invocable<F, Args...>::value) {
+  return detail::invoke(std::forward<F>(f), std::forward<Args>(args)...);
+}
+#endif
+
+// [func.memfn], member function adaptors
+
+// We only enable the C++17 implementation under C++2a here to account for
+// P0704: "Fixing const-qualified pointers to members" and LWG2210: "INVOKE-ing
+// a pointer to member with a `reference_wrapper` as the object expression".
+
+#if __cpp_lib_invoke /* C++17 */ && __cplusplus > 201703L /* C++2a */
+using std::mem_fn;
+#else
+template <typename R, typename T>
+typename detail::mem_fn_result<R T::*>::type mem_fn(R T::*pm) noexcept {
```

Comment:
We seem to use `T C::*` everywhere else aside from here. I guess it's `R T::*` here because that's
the signature used in the standard? I think we also have similar situations with `F, Args...` vs `Fn, ArgTypes...`.
I tend to ignore the ones used in the standard and use the names I like, personally. e.g., my `variant` is
declared `variant<Ts...>` rather than `variant<Types...>`. What are your thoughts?
`variant<Types...>`.

Reply:
I tend to just go with the names depicted in the standard for interfaces, but I don't particularly care.

Reply:
If you don't mind going with `T C::*`, and `F`/`Args`, let's go with that. I'd prefer to remember less signatures throughout.